votes/0606-pr-issue-association.md

```diff
@@ -0,0 +1,52 @@
+# A Vote for PR Issue Association
+
+## Proposal
+
+### Motivation
+
+To measure software evolution, we focus on features implemented, bugs fixed, and general enhancement. Issues document these items, while a pull request focuses on the change set and whether we implement it correctly.
+
+Normally, an issue documents a work item, and a pull request resolves it (despite cherry-picks). However, the relation between pull requests and issues is complicated in TiDB projects.
+
+- 1 issue with 0 PR: question, bug vanished without a reason, etc. 
+- 1 issue with 1 PR: normal case. 
+- 1 issue with n PRs: resolve a bug multiple times([example](https://github.com/pingcap/tidb/issues/24679)), or resolve a complex issue in steps([example](https://github.com/pingcap/tidb/issues/27652)). 
+- n issues with 1 PR: trade off resolving things as a whole (due to code coupling, or we can resolve multiple bugs at the common parent level)([example](https://github.com/pingcap/tidb/pull/27697)), duplicated issues. 
+- 0 issue with 1 PR: coding without a ticket([example](https://github.com/pingcap/tidb/pull/23022)). Even if the description is detailed or even if a design document is present, pull requests focus on the change set. It is about the implementation, not the question; if it is, we always mix discussion because the change set has a primary position. 
+
+This proposal tries to resolve the last case (0 issue with 1 PR), that a pull request must associate with issue(s). In this way,
+
+- Bring issues to a first-class development concept.
+- Benefit works tracking and contributors to join force in collaboration.
+- Ensure that all changes in release notes are associated with the corresponding issues.
+- Help the release manager and users understand software evolution.
+
+### Solution
+
+Issue number in PR body will become a mandatory requirement. Currently, the PR template contains one line `Issue Number: close #xxx`, but this is not mandatory. Contributors are free to leave it empty or delete the whole content. 
+
+After this proposal, the bot will check whether there is relevant issue number(s) in the PR body, by `Issue Number: ref #{issue-number}` or `Issue Number: close[sd]?|resolve[sd]?|fix(e[sd])? #{issue-number}`. `ref` means associated with an issue but does not close it. There should be no other content in the same line and the check is case-insensitive. Multiple issues should be separated by a comma, like `Issue Number: close #12341, close #12342`. For PRs trying to close issues in a different repository, contributors need to first create an issue in the same repository and use this issue to track. 
```

Comment:
can we tolerate “close https://github.com/pingcap/repo/issue/1234”? sometimes i like copy and paste urls and this also enables cross repo reference

Reply:
I think this suppose to work as long as it points to a valid issue link.

Reply:
Actually we follow Github's rule https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword with some tweaks. First add ref and then disallow cross repo reference. ref adds one semantic. And forcing the issue and PR in one repo helps other members like QA and release managers' work. We recommend create a new issue in the repo and use issue to track.

Reply:
> it helps other members like QA and release managers' work(they may need to apply labels to the issues)

if they need add labels, less issue means less manual work. if one problem is related to many repo (like need to update dependency) more redundant issues will be spawned.

Reply:
It makes some sense. But this means you have to take all related repos into considerations as a whole and switching back and forth to track the issue. If the repos don't share the same version strategy it also makes life harder. Updating dependencies sometimes means a bugfix so may affects multiple version so a dedicated issue could help.

Reply:
Talking about repos, we are consolidating repos. It's going to help reducing redundant issues.